RecoTauTag/HLTProducers/src/L2TauTagNNProducer.cc

```diff
@@ -566,71 +569,56 @@ void L2TauNNProducer::fillCaloRecHits(tensorflow::Tensor& cellGridMatrix,
-
-  std::vector<double> maxChi2_;
-  std::vector<double> pTSquaredSum(nv);
-
-  for (int j = nv - 1; j >= 0; --j) {
-    std::vector<int> trk_ass_to_vtx;
-    auto vtx_idx = patavtx_soa.sortInd[j];
-    assert(vtx_idx < nv);
-    for (int trk_idx = 0; trk_idx < maxTracks; trk_idx++) {
+                                                     std::vector<int>& trkGood,
+                                                     std::vector<int>& vtxGood) {
+   const auto maxTracks = patatracks_tsoa.stride();
+   const int nv = patavtx_soa.nvFinal;
+   trkGood.clear();
+   trkGood.reserve(maxTracks);
+   vtxGood.clear();
+   vtxGood.reserve(nv);
+   auto const* quality = patatracks_tsoa.qualityData();
+
+   // No need to sort either as the algorithms is just using the max (not even the location, just the max value of pt2sum).
+   std::vector<float> pTSquaredSum(nv,0);
+   std::vector<int> nTrkAssociated(nv,0);
+
+   for (int32_t trk_idx = 0; trk_idx < maxTracks; ++trk_idx) {
+     auto nHits = patatracks_tsoa.nHits(trk_idx);
+     if (nHits == 0){
+       break;
+     }
       int vtx_ass_to_track = patavtx_soa.idv[trk_idx];
-      if (vtx_ass_to_track == int16_t(vtx_idx))
-        trk_ass_to_vtx.push_back(trk_idx);
-    }
-    auto nt = trk_ass_to_vtx.size();
-    if (nt == 0) {
-      continue;
-    }
-    if (nt < 2) {
-      trk_ass_to_vtx.clear();
-      continue;
-    }
-    for (const auto& trk_idx : trk_ass_to_vtx) {
-      int vtx_ass_to_track = patavtx_soa.idv[trk_idx];
-      if (vtx_ass_to_track != vtx_idx)
-        continue;
-      double patatrackPt = patatracks_tsoa.pt[trk_idx];
-      if (patatrackPt < trackPtMin_)
-        continue;
-      if (patatracks_tsoa.chi2(trk_idx) > trackChi2Max_)
-        continue;
-      if (patatrackPt > trackPtMax_) {
-        patatrackPt = trackPtMax_;
-      }
-      pTSquaredSum.at(vtx_idx) += patatrackPt * patatrackPt;
-    }
-  }
-  std::vector<size_t> sortIdxs(nv);
-  std::iota(sortIdxs.begin(), sortIdxs.end(), 0);
-  std::sort(sortIdxs.begin(), sortIdxs.end(), [&](size_t const i1, size_t const i2) {
-    return pTSquaredSum[i1] > pTSquaredSum[i2];
-  });
-  auto const minFOM_fromFrac = pTSquaredSum[sortIdxs.front()] * fractionSumPt2_;
-
-  for (int j = nv - 1; j >= 0; --j) {
-    auto idx = patavtx_soa.sortInd[j];
-
-    if (VtxGood.size() >= maxVtx_) {
-      break;
-    }
-    if (pTSquaredSum[idx] >= minFOM_fromFrac && pTSquaredSum[idx] > minSumPt2_) {
-      VtxGood.push_back(idx);
-    }
-  }
-  return VtxGood;
+      if(vtx_ass_to_track>=0 && vtx_ass_to_track< nv){
+        auto patatrackPt = patatracks_tsoa.pt[trk_idx];
+        ++nTrkAssociated[vtx_ass_to_track];
+         if (patatrackPt >= trackPtMin_ && patatracks_tsoa.chi2(trk_idx)<=trackChi2Max_){
+           patatrackPt = std::min(patatrackPt, trackPtMax_);
+           pTSquaredSum[vtx_ass_to_track] += patatrackPt * patatrackPt;
+         }
+       }
+     auto q = quality[trk_idx];
+     if (nHits > 0 && q >= pixelTrack::Quality::loose){
```

Comment:
```suggestion
     if (nHits > 0 and quality[trk_idx] >= pixelTrack::Quality::loose){
```
I don't see the use of `q` in this case.